inst/app/modules/main_map.R

```diff
@@ -9,12 +27,45 @@ hk_accidents_valid_sf <- st_as_sf(x = hk_accidents_valid, coords = c("longitude"
 # TODO: investigate why
 hk_accidents_valid_sf$Date <- as.Date(hk_accidents_valid_sf$Date, format = "%Y-%m-%d")
 
-CARTODB_POSITRON_TILE_URL <- "https://cartodb-basemaps-{s}.global.ssl.fastly.net/light_all/{z}/{x}/{y}.png"
+# Combine date and time together as a complete POSIXct class time column
+# Easier for formatting
+hk_accidents_valid_sf$Date_Time <- as.POSIXct(
+  strptime(
+    paste0(hk_accidents_valid_sf$Date, " ", hk_accidents_valid_sf$Time),
+    format = "%Y-%m-%d %H%M",
+    tz = "Asia/Hong_Kong"
+    )
+  )
+
 
 output$main_map <- renderLeaflet({
-  leaflet() %>%
-    addTiles(urlTemplate = CARTODB_POSITRON_TILE_URL) %>%
-    setView(lng = 114.2, lat = 22.3, zoom = 12)
+  overview_map <- leaflet() %>%
+    setView(lng = 114.2, lat = 22.3, zoom = 12) %>%
+    # Add geocoder map widget
+    addSearchOSM(options = searchOptions(hideMarkerOnCollapse = TRUE))
+
+  # Subset basemap providers to be used for the map
+  SELECTED_BASEMAPS <- leaflet::providers[c("CartoDB.Positron", "Stamen.TonerLite", "OpenStreetMap")]
+
+  # Add the basemap tiles in background
+  # http://rstudio.github.io/leaflet/morefeatures.html
+  for (provider in SELECTED_BASEMAPS) {
+    overview_map <- addProviderTiles(overview_map, provider, group = provider)
+  }
+
+  # Add change basemap widget
+  overview_map %>%
+    addLayersControl(baseGroups = names(basemap_providers),
+                     options = layersControlOptions(collapsed = TRUE),
+                     position = "topleft") %>%
+    htmlwidgets::onRender("
+    function(el, x) {
+      var myMap = this;
+      myMap.on('baselayerchange',
+        function (e) {
+          myMap.minimap.changeLayer(L.tileLayer.provider(e.name));
+        })
+    }")
 })
```

Comment:
Looks like you're introducing JavaScript (?) code here and you're piping it to some R code. Is this best practice? If there's no alternative ways to do this, no worries, just looks odd to have a mix of R and JavaScript.

Reply:
After further reading into the code, I find this section redundant as this is used for changing the basemap of the MiniMap. The `addLayersControl` function above already provides functionality to change the basemap of the main map. (Sorry that I directly copy-pasted the code from the [manual of leaflet](https://rstudio.github.io/leaflet/morefeatures.html))

Still, in terms of the function itself, the **htmlwidgets::onRender** is designed to wrap JS code into various HTML widgets created (See the [documentation](https://www.rdocumentation.org/packages/htmlwidgets/versions/1.5.3/topics/onRender)), with leaflet being no exception. I also find it odd to mix R and JS together, yet seems it is the easiest way to customise the web widgets with R.

Reply:
No worries! I did think there was a native leaflet function that enabled the changing of the basemap. 

On `htmlwidgets::onRender()`, I was thinking more along the lines of whether it is best practice to specify a separate JavaScript file that holds this code then call that into the function. Thus, when reviewing the code, the reviewer can clearly see which parts are R and which parts are JS.

Reply:
In line with the idea of separating the JS file. We may check if `htmlwidgets::onRender()` support reading external files or if there are any possible workarounds.